networkx/utils/decorators.py

```diff
@@ -122,140 +123,93 @@ def open_file(path_arg, mode="r"):
     --------
     Decorate functions like this::
 
-       @open_file(0,'r')
+       @open_file(0,"r")
        def read_function(pathname):
            pass
 
-       @open_file(1,'w')
-       def write_function(G,pathname):
+       @open_file(1,"w")
+       def write_function(G, pathname):
+           pass
+
+       @open_file(1,"w")
+       def write_function(G, pathname="graph.dot")
            pass
 
-       @open_file(1,'w')
-       def write_function(G, pathname='graph.dot')
+       @open_file("pathname","w")
+       def write_function(G, pathname="graph.dot")
            pass
 
-       @open_file('path', 'w+')
+       @open_file("path", "w+")
        def another_function(arg, **kwargs):
-           path = kwargs['path']
+           path = kwargs["path"]
            pass
+
+    Notes
+    -----
+    Note that this decorator solves the problem when a path argument is
+    specified as a string, but it does not handle the situation when the
+    function wants to accept a default of None (and then handle it).
+
+    Here is an example of how to handle this case:
+
+      @open_file("path")
+      def some_function(arg1, arg2, path=None):
```

Comment:
I think this alternative creates a single tempfile that gets reused rather than constructing a new one for each call of the function.  But I think the example could avoid the `close_fobj` flag by just testing `if path is None:` at the end instead of testing `close_fobj`.  But I'm not it makes much difference -- does the logic flow better?  maybe I little...  I'll make that change to the example.  

Check this to see if it gets better.

Reply:
This is fine.  Again, I don't think the check is necessary.  The file can simply be closed.  The real logic you want to show is that you get to do something special if the input is None.

I like the `default=` API better, I think, but that's for you and @boothby to decide.

Reply:
I think we want to leave the file open when it has been opened by the caller. They may not be done with it.
We do want to close it when the decorator opened it.  But we don't want to close everything we use.

Reply:
All our decorator does is close it, but I guess as a "general principle"?

Reply:
No -- our `open_file` sometimes returns `lambda: None` for the second return value (the try_finally function).  If path is a string or a pathlib instance we provide the close function.
Otherwise, `path` is probably `None`, or a file handle so we `return path, lambda: None`  (don't close).

Reply:
Aha, thanks, I didn't realize that!

Reply:
The more I think about it, I think this decorator should go.

Having a `with nx_open(x)` is as short, less confusing, and more explicit.

Reply:
But doesn’t that close the file handle?  (…or rather, what is the way to avoid closing it while using “with”?)

Reply:
If you want to deprecate that function, please bring it up in another issue.  We'll need to retain this functionality until at least 3.0, so I won't be able to simplify `argmap` until then.

Reply:
I meant the standard straightforward way, without any decoration:

```
with nx_open(filename):
    ... do the work ...
```

And `nx_open` can open a file, create temporaries, whatever.

@boothby No problem holding off on a refactor like that until 3.0, if everyone agrees it's a good step.